File: tesla-energy-project/frontend/src/types/device.ts
export enum DeviceType {
  POWERWALL = "powerwall",
  SOLAR_INVERTER = "solar_inverter",
  WALL_CONNECTOR = "wall_connector",
}

export enum DeviceStatus {
  ONLINE = "online",
  OFFLINE = "offline",
  STANDBY = "standby",
  CHARGING = "charging",
  DISCHARGING = "discharging",
  FAULT = "fault",
  UPDATING = "updating",
}

export enum OperationMode {
  BACKUP = "backup",
  SELF_POWERED = "self_powered",
  TIME_BASED_CONTROL = "time_based_control",
  ADVANCED = "advanced",
}

export enum ControlCommandType {
  CHARGE_NOW = "charge_now",
  STOP_CHARGING = "stop_charging",
  ISOLATE_FROM_GRID = "isolate_from_grid",
  REJOIN_GRID = "rejoin_grid",
  REBOOT = "reboot",
  FIRMWARE_UPDATE = "firmware_update",
  SET_BACKUP_RESERVE = "set_backup_reserve",
  SET_OPERATION_MODE = "set_operation_mode",
}

export interface DeviceConfig {
  backup_reserve_percent: number;
  operation_mode: OperationMode;
  grid_charging_enabled: boolean;
  storm_watch_enabled: boolean;
  time_of_use_enabled: boolean;
  firmware_version: string;
  installation_date?: string;
}

export interface Device {
  device_id: string;
  serial_number: string;
  device_type: DeviceType;
  model: string;
  firmware_version: string;
  status: DeviceStatus;
  config: DeviceConfig;
  location?: string;
  installed_at: string;
  last_seen: string;
  battery_capacity_kwh: number;
  solar_capacity_kw?: number;
}

export interface Telemetry {
  device_id: string;
  timestamp: string;
  battery_charge_percent: number;
  battery_power_kw: number;
  solar_power_kw: number;
  grid_power_kw: number;
  home_power_kw: number;
  battery_temperature_c: number;
  inverter_temperature_c?: number;
  voltage?: number;
  frequency_hz?: number;
  state_of_health: number;
  cycles?: number;
  backup_reserve_percent: number;
  operation_mode: string;
}

export interface ControlCommand {
  command: ControlCommandType;
  parameters?: Record<string, any>;
  timestamp?: string;
}

export interface HealthAnalysis {
  device_id: string;
  overall_health_score: number;
  analysis: string;
  recommendations: string[];
  key_metrics: Record<string, any>;
  timestamp: string;
}

export interface DeviceListResponse {
  devices: Device[];
  total: number;
  page: number;
  page_size: number;
}

File: tesla-energy-project/backend/app/models/device.py
from datetime import datetime
from enum import Enum
from typing import Optional, List
from pydantic import BaseModel, Field


class DeviceType(str, Enum):
    POWERWALL = "powerwall"
    SOLAR_INVERTER = "solar_inverter"
    WALL_CONNECTOR = "wall_connector"


class OperationMode(str, Enum):
    BACKUP = "backup"
    SELF_POWERED = "self_powered"
    TIME_BASED_CONTROL = "time_based_control"
    ADVANCED = "advanced"


class DeviceStatus(str, Enum):
    ONLINE = "online"
    OFFLINE = "offline"
    STANDBY = "standby"
    CHARGING = "charging"
    DISCHARGING = "discharging"
    FAULT = "fault"
    UPDATING = "updating"


class ControlCommandType(str, Enum):
    CHARGE_NOW = "charge_now"
    STOP_CHARGING = "stop_charging"
    ISOLATE_FROM_GRID = "isolate_from_grid"
    REJOIN_GRID = "rejoin_grid"
    REBOOT = "reboot"
    FIRMWARE_UPDATE = "firmware_update"
    SET_BACKUP_RESERVE = "set_backup_reserve"
    SET_OPERATION_MODE = "set_operation_mode"


class DeviceConfig(BaseModel):
    backup_reserve_percent: float = Field(default=20.0, ge=0.0, le=100.0)
    operation_mode: OperationMode = OperationMode.SELF_POWERED
    grid_charging_enabled: bool = False
    storm_watch_enabled: bool = True
    time_of_use_enabled: bool = False
    firmware_version: str = "23.44.1"
    installation_date: Optional[datetime] = None


class ControlCommand(BaseModel):
    command: ControlCommandType
    parameters: Optional[dict] = None
    timestamp: Optional[datetime] = Field(default_factory=datetime.utcnow)


class Device(BaseModel):
    device_id: str
    serial_number: str
    device_type: DeviceType
    model: str
    firmware_version: str
    status: DeviceStatus
    config: DeviceConfig
    location: Optional[str] = None
    installed_at: datetime
    last_seen: datetime
    battery_capacity_kwh: float = 13.5  # Powerwall 2 capacity
    solar_capacity_kw: Optional[float] = None
    
    class Config:
        json_schema_extra = {
            "example": {
                "device_id": "PW-001-ABC123",
                "serial_number": "SN123456789",
                "device_type": "powerwall",
                "model": "Powerwall 2",
                "firmware_version": "23.44.1",
                "status": "online",
                "config": {
                    "backup_reserve_percent": 20.0,
                    "operation_mode": "self_powered",
                    "grid_charging_enabled": False,
                    "storm_watch_enabled": True
                },
                "location": "Garage",
                "installed_at": "2023-01-15T10:00:00Z",
                "last_seen": "2024-01-20T14:30:00Z",
                "battery_capacity_kwh": 13.5,
                "solar_capacity_kw": 8.5
            }
        }


class DeviceListResponse(BaseModel):
    devices: List[Device]
    total: int
    page: int = 1
    page_size: int = 50

